contracts/command-batcher/src/contract.rs

```diff
@@ -1,44 +1,343 @@
+        }
+
+        let snapshot = snapshot(deps.querier, env.block, &config)?;
+
+        for participant in snapshot.participants.keys() {
+            if !pub_keys.contains_key(participant) {
+                return Err(ContractError::PublicKeyNotFound {
+                    participant: participant.to_owned(),
+                });
+            }
+        }
+
+        let keygen_msg = multisig::msg::ExecuteMsg::KeyGen {
+            key_id: config.service_name,
+            snapshot,
+            pub_keys,
+        };
+
+        Ok(Response::new().add_message(WasmMsg::Execute {
+            contract_addr: config.multisig.into(),
+            msg: to_binary(&keygen_msg)?,
+            funds: vec![],
+        }))
+    }
+
+    fn snapshot(
+        querier: QuerierWrapper,
+        block: BlockInfo,
+        config: &Config,
+    ) -> Result<Snapshot, ContractError> {
+        let query_msg = service_registry::msg::QueryMsg::GetActiveWorkers {
+            service_name: config.service_name.clone(),
+            chain_name: config.chain_name.to_string(),
+        };
+
+        let active_workers: Vec<Worker> = querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: config.service_registry.to_string(),
+            msg: to_binary(&query_msg)?,
+        }))?;
+
+        let participants = active_workers
+            .into_iter()
+            .map(Worker::try_into)
+            .collect::<Result<Vec<Participant>, service_registry::ContractError>>()
+            .map_err(
+                |err: service_registry::ContractError| ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                },
+            )?
+            .try_into()
+            .map_err(|err: axelar_wasm_std::nonempty::Error| {
+                ContractError::InvalidParticipants {
+                    reason: err.to_string(),
+                }
+            })?;
+
+        Ok(Snapshot::new(
+            block
+                .time
+                .try_into()
+                .expect("violated invariant: block time is zero"),
+            block
+                .height
+                .try_into()
+                .expect("violated invariant: block height is zero"),
+            config.signing_threshold,
+            participants,
+        ))
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, reply: Reply) -> Result<Response, ContractError> {
+    let reply_id = reply.id;
+
+    match parse_reply_execute_data(reply) {
+        Ok(MsgExecuteContractResponse { data: Some(data) }) => {
+            let command_batch_id = REPLY_ID_TO_BATCH.load(deps.storage, reply_id)?;
+
+            let session_id =
+                from_binary(&data).map_err(|_| ContractError::InvalidContractReply {
+                    reason: "invalid multisig session ID".to_string(),
+                })?;
+
+            let proof_id = ProofID::new(&command_batch_id, &session_id);
```

Comment:
Can we make the proof_id deterministic, so we don't have to process the reply? It could be a pair of `(key_id, msg_to_sign)`

Reply:
This would require changes to the multisig contract, since session_id is just a counter, so I think out of scope for this PR. But it would be nice for the future if the multisig contract derived session_id from the input (msg to sign, and key_id). Then we don't need to actually process the reply from the multisig contract here, and avoid the complexity associated with that.

In general, I'd like to see this whole batch creation and signing flow become idempotent.

Reply:
Agree this should be out of scope for now. Before we implement deterministic id for the multisig we should implement deletion of failed sessions (completed too?), otherwise I think we won't be able to retry multisig for same batch

Reply:
I don't think we need to delete completed sessions, but yes we should have some way of restarting a failed session

Reply:
task to remove proof_id: https://axelarnetwork.atlassian.net/browse/AXE-1514